ocs_ci/deployment/flexy.py

```diff
@@ -0,0 +1,274 @@
+            )
+            if cp.returncode:
+                if cp.stderr:
+                    logger.error(cp.stderr)
+                raise exceptions.CommandFailed("Flexy run failed")
+            logger.info("Flexy run finished successfully")
+            self.exitThread = True
+        for res in TimeoutSampler(60, 1, future.done):
+            if not res:
+                logger.info("Waiting for logger thread to join")
+
+    def build_install_cmd(self):
+        """
+        Build flexy command line for 'deploy' operation
+
+        """
+        cmd = "sudo podman run --rm=true"
+        flexy_container_args = self.build_container_args()
+        cmd_string = " ".join([cmd, " ".join(flexy_container_args)])
+        return cmd_string
+
+    def build_destroy_cmd(self):
+        """
+        Build flexy command line for 'destroy' operation
+
+        """
+        cmd = "sudo podman run --rm=true"
+        flexy_container_args = self.build_container_args()
+        cmd_string = " ".join([cmd, " ".join(flexy_container_args), 'destroy'])
+        return cmd_string
+
+    def get_log(self):
+        """
+        This function runs in a seperate thread and continuously reads
+        logs from flexy podman
+
+        """
+        for res in TimeoutSampler(20, 1, os.path.exists, self.flexy_log_file):
+            if res:
+                break
+        log_fd = open(self.flexy_log_file, "r")
+
+        while True:
+            line = log_fd.readline()
+            if line:
+                logger.info(line.strip())
+            if self.exitThread:
+                break
+        logger.info("Logger thread done printing flexy logs")
+        # exit only this thread
+        sys.exit()
+
+    def build_container_args(self):
+        """
+        Builds most commonly used arguments for flexy container
+
+        """
+        args = list()
+        args.append(f"--log-opt={self.flexy_log_file}")
+        args.append(f"--env-file={self.flex_env_file}")
+        args.append(f"-w={self.flexy_mnt_container_dir}")
+        args.append(
+            f"--mount=type=bind,source={self.flexy_mnt_host_dir},"
+            f"destination=f{self.flexy_mnt_container_dir},relabel=shared"
+        )
+        args.append(
+            f"--mount=type=bind,source={self.flexy_host_private_conf_dir_path},"
+            f"destination={self.flexy_mnt_private_conf_dir},relabel=shared"
+        )
+        args.append(f"{self.flexy_img_url}")
+        return args
+
+    def clone_and_unlock_ocs_private_conf(self):
+        """
+        Clone ocs_private_conf (flexy env and config) repo into
+        flexy_mnt_host_dir
+
+        """
+        clone_repo(self.flexy_private_conf_url, self.flexy_host_private_conf_dir_path)
+        # git-crypt unlock /path/to/keyfile
+        old_cwd = os.getcwd()
+        os.chdir(self.flexy_host_private_conf_dir_path)
+        cp = subprocess.run(
+            f'git-crypt unlock {constants.FLEXY_GIT_CRYPT_KEYFILE}',
```

Comment:
ack

Reply:
after discussion with @dahorak we were thinking about approach to use this variable for FLEXY_GIT_CRYPT_KEYFILE from config if defined which will allow the user to specify key file on it's own via config file. If it will not be defined it will use the gpg key. So something like:
`f'git-crypt unlock {config.RUN.get('FLEXY_GIT_CRYPT_KEYFILE', '')}',`

Reply:
@dahorak @petr-balogh 
This one is still pending, I am bit confused about this crypt file requirement,
we are having 3 options to specify flexy env 

1.Directly pass env file's absolute path to ocs-ci
2. Pass private-conf repo through yaml to ocs-ci
3. If nothing is specified we will go with default flexy-ocs-private conf repo.

In all the above cases we expect crypt key to be present in env file so that we don't have to handle it in ocs-ci ?

WDYT ?

Reply:
> @dahorak @petr-balogh
> This one is still pending, I am bit confused about this crypt file requirement,
> we are having 3 options to specify flexy env
> 
> 1.Directly pass env file's absolute path to ocs-ci
If you pass absolute path you expect that this file is not encrypted and is fully readable and contains key inside. (This case you don't clone the repo - so no need for unlock)
> 2. Pass private-conf repo through yaml to ocs-ci
If you pass the repository then ocs-ci will clone it, but to be able to pass any env file from this repo you have to unlock the repository. So this case the repo should be unlocked by `FLEXY_GIT_CRYPT_KEYFILE` - which we will generate to jenkins config file or can be done by CLI parameter.   If `FLEXY_GIT_CRYPT_KEYFILE` not passed - we will try to unlock with GPG key so the only command `git-crypt unlock` will be executed (and expect user to have gpg key)
> 3. If nothing is specified we will go with default flexy-ocs-private conf repo.
This will be the same approach as above.
> 
> In all the above cases we expect crypt key to be present in env file so that we don't have to handle it in ocs-ci ?
As you need to unlock the repository before you pass env file to flexy you need to handle unlock in ocs-ci.
> 
> WDYT ?

Reply:
> > @dahorak @petr-balogh
> > This one is still pending, I am bit confused about this crypt file requirement,
> > we are having 3 options to specify flexy env
> > 1.Directly pass env file's absolute path to ocs-ci
> > If you pass absolute path you expect that this file is not encrypted and is fully readable and contains key inside. (This case you don't clone the repo - so no need for unlock)
> > 2. Pass private-conf repo through yaml to ocs-ci
> > If you pass the repository then ocs-ci will clone it, but to be able to pass any env file from this repo you have to unlock the repository. So this case the repo should be unlocked by `FLEXY_GIT_CRYPT_KEYFILE` - which we will generate to jenkins config file or can be done by CLI parameter.   If `FLEXY_GIT_CRYPT_KEYFILE` not passed - we will try to unlock with GPG key so the only command `git-crypt unlock` will be executed (and expect user to have gpg key)

@petr-balogh @dahorak  ,
In this case where user wants to pass his/her own private-conf repo How we want to pass location of keyfile ? through URL on Jenkin's job page ? OR a facility to upload keyfile in the Jenkin's job file ?

> > 3. If nothing is specified we will go with default flexy-ocs-private conf repo.
> > This will be the same approach as above.
> > In all the above cases we expect crypt key to be present in env file so that we don't have to handle it in ocs-ci ?
> > As you need to unlock the repository before you pass env file to flexy you need to handle unlock in ocs-ci.
> > WDYT ?

Reply:
@shylesh from jenkins side we will use jenkins credentials while we are executing ocs-ci and will pass this location to the ocs-ci in way how you will integrate it CLI pointing to the file location or will update config file we generate in jenkins to provide this in proper section in conf file.

Do not think about it much, for now we will not use any other keyfile than only one which we have in our current private repo - do not expect that user will change that in their own fork for now.

Once it will be needed we will manage some way we will agree on but for now it should be passed from jenkins credentials like we use already with auth.yaml file and so on.